packages/engine/crawlers/solana/transactionHistory.ts

```diff
@@ -0,0 +1,530 @@
+		searchTransactionHistory: true,
+	});
+	return result.value?.confirmationStatus;
+};
+
+const isAbstractionFee = (meta: ParsedTransactionWithMeta['meta']) => {
+	const tokenBalances = meta?.postTokenBalances;
+
+	for (const tokenBalance of tokenBalances ?? []) {
+		if (tokenBalance.owner === FEE_PAYER) {
+			return true;
+		}
+	}
+
+	return false;
+};
+
+const getTransactionTokenForFee = async (
+	instruction: ParsedInstruction | PartiallyDecodedInstruction,
+	meta: ParsedTransactionWithMeta['meta'],
+	solanaContext: SolanaContext,
+) => {
+	if (isAbstractionFee(meta)) {
+		const tokenBalances = meta?.postTokenBalances;
+
+		for (const tokenBalance of tokenBalances ?? []) {
+			if (!tokenBalance.owner) continue;
+
+			if (tokenBalance.owner === FEE_PAYER) {
+				const network = Networks.solana;
+				const account: TokenAccount = {
+					mint: tokenBalance.mint,
+					owner: tokenBalance.owner,
+					address: (instruction as ParsedInstruction).parsed?.info.source,
+					balance: tokenBalance.uiTokenAmount.uiAmount?.toString() || '',
+					decimals: tokenBalance.uiTokenAmount.decimals,
+				};
+				const metadata = await getMetadata(solanaContext, tokenBalance.mint);
+
+				const token: Token = {
+					network,
+					account,
+					metadata,
+				};
+
+				return token;
+			}
+		}
+	}
+
+	return {
+		network: Networks.solana,
+		account: {
+			mint: solMint,
+			owner: 'system',
+			address: solMint,
+			balance: '',
+			decimals: 9,
+		},
+		metadata: await getMetadata(solanaContext, solMint),
+	};
+};
+
+const getTransactionFee = (
+	transaction: ParsedTransactionWithMeta,
+	tokenForFee: Token,
+) => {
+	let fee = (transaction.meta?.fee ?? 5000) / 10 ** 9;
+
+	if (isAbstractionFee(transaction.meta)) {
+		const instruction = transaction.transaction.message.instructions[0];
+		const feeLamports = (instruction as ParsedInstruction).parsed?.info.amount;
+		const decimals = tokenForFee.account.decimals;
+
+		if (feeLamports) {
+			fee = feeLamports / 10 ** decimals;
+		}
+	}
+
+	return fee;
+};
+
+const getTransactionToken = async (
+	instruction: ParsedInstruction | PartiallyDecodedInstruction,
+	meta: ParsedTransactionWithMeta['meta'],
+	type: 'sent' | 'received',
+	sender: string,
+	receiver: string,
+	solanaContext: SolanaContext,
+) => {
+	let token: Token = {
+		network: Networks.solana,
+		account: {
+			mint: '',
+			owner: '',
+			address: '',
+			balance: '',
+			decimals: 0,
+		},
+		metadata: {
+			name: 'Unknown',
+			symbol: '',
+			imageUri: '/img/send-token/unknown-token.jpeg',
+		},
+	};
+
+	if (
+		!meta?.preTokenBalances ||
+		!meta?.postTokenBalances ||
+		meta?.postTokenBalances.length === 0 ||
+		meta?.preTokenBalances.length === 0
+	) {
+		const solMetadata = await getMetadata(solanaContext, solMint);
+
+		token = {
+			network: Networks.solana,
+			account: {
+				mint: solMint,
+				owner: '',
+				address: '',
+				balance: '',
+				decimals: 9,
+			},
+			metadata: solMetadata,
+		};
```

Comment:
if  you have already initialized `token`, just assign what changed, not reassign to new object